src/buildblock/GeometryBlocksOnCylindrical.cxx

```diff
@@ -0,0 +1,301 @@
+                          );
+}
+
+
+Succeeded
+GeometryBlocksOnCylindrical::
+find_detection_position_given_cartesian_coordinate(DetectionPosition<>& det_pos,
+                                                   const CartesianCoordinate3D<float>& cart_coord) const
+{
+  /*! first round the cartesian coordinates, it might happen that the cart_coord
+   is not precisely pointing to the center of the crystal and
+   then the det_pos cannot be found using the map
+  */
+  //rounding cart_coord to 3 decimal place and find det_pos
+  CartesianCoordinate3D<float> rounded_cart_coord;
+  rounded_cart_coord.z() = round(cart_coord.z()*1000.0F)/1000.0F;
+  rounded_cart_coord.y() = round(cart_coord.y()*1000.0F)/1000.0F;
+  rounded_cart_coord.x() = round(cart_coord.x()*1000.0F)/1000.0F;
+	if (detection_position_map_given_cartesian_coord_keys_3_decimal.count(rounded_cart_coord))
+	{
+          det_pos =	detection_position_map_given_cartesian_coord_keys_3_decimal.at(rounded_cart_coord);
+          return Succeeded::yes;
+	}
+	else
+	{
+          //rounding cart_coord to 2 decimal place and find det_pos
+		rounded_cart_coord.z() = round(cart_coord.z()*100.0F)/100.0F;
+		rounded_cart_coord.y() = round(cart_coord.y()*100.0f)/100.0F;
+		rounded_cart_coord.x() = round(cart_coord.x()*100.0F)/100.0F;
+		if (detection_position_map_given_cartesian_coord_keys_2_decimal.count(rounded_cart_coord))
+		{
+                  det_pos =	detection_position_map_given_cartesian_coord_keys_2_decimal.at(rounded_cart_coord);
+                  return Succeeded::yes;
+		}
+		else
+		{
+			warning("cartesian coordinate (x, y, z)=(%f, %f, %f) does not exist in the inner map",
+							cart_coord.x(), cart_coord.y(), cart_coord.z());
+			return Succeeded::no;
+		}
+	}
+}
+
+void
+GeometryBlocksOnCylindrical::
+build_crystal_maps()
+{
+	// local variables to describe scanner
+	int num_axial_crystals_per_block = get_scanner_ptr()->get_num_axial_crystals_per_block();
+	int num_transaxial_crystals_per_block = get_scanner_ptr()->get_num_transaxial_crystals_per_block();
+	int num_axial_blocks = get_scanner_ptr()->get_num_axial_blocks();
+	int num_transaxial_blocks_per_bucket = get_scanner_ptr()->get_num_transaxial_blocks_per_bucket();
+	int num_transaxial_buckets = get_scanner_ptr()->get_num_transaxial_blocks()/num_transaxial_blocks_per_bucket;
+	int num_detectors_per_ring = get_scanner_ptr()->get_num_detectors_per_ring();
+	float axial_block_spacing = get_scanner_ptr()->get_axial_block_spacing();
+	float transaxial_block_spacing = get_scanner_ptr()->get_transaxial_block_spacing();
+	float axial_crystal_spacing = get_scanner_ptr()->get_axial_crystal_spacing();
+	float transaxial_crystal_spacing = get_scanner_ptr()->get_transaxial_crystal_spacing();
+	std::string scanner_orientation = get_scanner_ptr()->get_scanner_orientation();
+
+	// check for the scanner orientation
+	if (scanner_orientation=="Y" || num_transaxial_buckets%4==0 )
```

Comment:
If number of buckets in transaxial direction is a multiple of 4 this causes scanner orientation to alway to Y.

Reply:
That's true. When  `num_transaxial_buckets%4==0`, there's always two faces of the scanner perpendicular to the y axis.

Reply:
not necessarily since we can have y-axis going though two corners of the scanner rather than two faces. Due to this condition it will not allow us to get orientation X and num_transaxial_buckets%4==0.

Reply:
You can have infinite number of orientations of the scanner toward x and y axis. Not all of them are defined here. Only two common cases are defined: the polygon face is either perpendicular to x axis or to y axis.
If `num_transaxial_buckets%4==0` is true and the y axis goes through the corners of the scanner the x-axis goes through the corner too. It means it is none of the defined orientations.